matchengine/engine.py

```diff
@@ -219,103 +219,219 @@ def create_trial_tree(self, raw_data, no_validate=False):
+            'TRUE_HUGO_SYMBOL': 1,
+            'ONCOKB_VARIANT': 1,
+            '_id': 1
+        }
+
+        matched_hugo_results = list(self.db.genomic.find(query, proj))
+        for genomic_item in matched_hugo_results:
+            if "ONCOKB_VARIANT" in genomic_item and genomic_item["ONCOKB_VARIANT"] and oncokb_variant in genomic_item["ONCOKB_VARIANT"]:
+                results.append(genomic_item)
+
+        if hugo_symbol_negative or oncokb_variant_negative:
+            matched_sample_ids = self.all_match - set(x['SAMPLE_ID']for x in results)
+            # add genomic alterations per sample id
+            matched_genomic_info = [{
+                'sample_id': sample_id,
+                'match_type': 'oncokb_variant',
+                'genomic_alteration': "!" + hugo_symbol + " " + "!" + oncokb_variant
+            } for sample_id in matched_sample_ids]
+        else:
+            matched_sample_ids = set(item['SAMPLE_ID'] for item in results)
+
+            for item in results:
+                genomic_info = {
+                    'match_type': 'oncokb_variant',
+                    'genomic_alteration': hugo_symbol + " " + oncokb_variant
+                }
+                for field in proj:
+                    if field in item:
+                        if field == '_id':
+                            genomic_info['genomic_id'] = item[field]
+                        else:
+                            genomic_info[field.lower()] = item[field]
+                matched_genomic_info.append(genomic_info)
+
+        return matched_sample_ids, matched_genomic_info
+
+    def run_query(self, node):
+        """
+        Runs genomic or clinical query against Mongo database and returns a set of sample ids that matched
+
+        :param node: node location with the trial match tree
+        :param db: database connection
+
+        :returns
+            matched_sample_ids: set of matched sample ids
+            matched_genomic_info: genomic information regarding each match
+        """
+
+        matched_genomic_info = []
+        matched_sample_ids = set()
+        run_oncokb_match = False
 
-                        # add unique matches by sample id
-                        matched_genomic_info.append(genomic_info)
+        # execute query against genomic table
+        if node['type'] == 'genomic':
+            item = node['value']
+
+            # check if "item" can run general_match(). "hugo_symbol" is required for any match method
+            general_map_keys = ["variant_category", "protein_change", "wildcard_protein_change",
+                    "variant_classification", "exon", "cnv_call", "wildtype", "mmr_status", "ms_status"]
 
-                    matched_sample_ids = set(item['SAMPLE_ID'] for item in results)
+            # The match method of thrid party plugins should be run before general_match()
+            # since general_match() will delete map key "oncokb_variant".
+            if 'oncokb_variant' in item:
```

Comment:
```
for key in item.keys():
    if key in general_map_keys and 'hugo_symbol' in item:
            matched_sample_ids, matched_genomic_info = self.general_match(item)

# The match method of thrid party plugins should be run before general_match()
# since general_match() will delete map key "oncokb_variant".
if 'oncokb_variant' in item:
    oncokb_matched_sample_ids, oncokb_matched_genomic_info = self.oncokb_match(item)
    if matched_sample_ids is Array:
        matched_sample_ids.intersection(oncokb_matched_sample_ids)
        if(len(matched_sample_ids) > 0)
            for genomic_info in oncokb_matched_genomic_info:
                if genomic_info['sample_id'] in matched_sample_ids:
                    temp_matched_genomic_info.append(genomic_info)
            matched_genomic_info = temp_matched_genomic_info
    else:
        matched_sample_ids = oncokb_matched_sample_ids
```

Reply:
For `if matched_sample_ids is Array:`, I think we need to set a `run_general_match` flag. `matched_sample_ids` has been initialized at the beginning of run_query(), so you cannot know whether general_match() has been run through check type or length of `matched_sample_ids`.